File: HackerRank/diagonalDifference/difference.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.replace(/\s*$/, '')
        .split('\n')
        .map(str => str.replace(/\s*$/, ''));

    main();
});

function readLine() {
    return inputString[currentLine++];
}

function rightDiagonal(arrays) {
    let sum = 0;
    for (var index in arrays) {
        sum += arrays[index][index];
    }
    return sum;
}

function leftDiagonal(arrays){
    let sum    = 0;
    let row    = 0;
    let column = arrays.length - 1;
    while (row < arrays.length) {
        sum += arrays[row][column];
        row++;
        column--;
    }
    return sum;
}

// Complete the diagonalDifference function below.
function diagonalDifference(arrays) {
    let diff = (rightDiagonal(arrays) - leftDiagonal(arrays));
    return Math.abs(diff);
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    console.log("How many numbers wide/tall is the grid?");
    const n = parseInt(readLine(), 10);

    let a = Array(n);

    for (let i = 0; i < n; i++) {
        console.log(`Input Line #${i}.`);
        a[i] = readLine().split(' ').map(aTemp => parseInt(aTemp, 10));
    }

    let result = diagonalDifference(a);

    ws.write(result + "\n");

    ws.end();
}

File: HackerRank/diagonalDifference/difference.rb
#!/bin/ruby

require 'json'
require 'stringio'

def right_diagonal_sum(arrays)
    outsum = 0
    arrays.each_index do |index|
        outsum += arrays[index][index]
    end
    return outsum
end

def left_diagonal_sum(arrays)
    outsum = 0
    row = 0
    column = arrays.length - 1
    while row < arrays.length do
        puts "Sum is currently #{outsum}, at row #{row} and column #{column}, so adding #{arrays[row][column]}"
        theRow = arrays[row]
        puts "theRow[column] = #{theRow[column]}"
        outsum += theRow[column]
        column -= 1
        row    += 1
    end
    return outsum
end

# Complete the diagonalDifference function below.
def diagonalDifference(a)
    diff = right_diagonal_sum(a) - left_diagonal_sum(a)
    return diff.abs
end

fptr = File.open(ENV['OUTPUT_PATH'], 'w')

n = gets.to_i

a = Array.new(n)

n.times do |i|
    a[i] = gets.rstrip.split(' ').map(&:to_i)
end

result = diagonalDifference a

fptr.write result
fptr.write "\n"

fptr.close()
